src/System.Net.Security/tests/FunctionalTests/SslStreamSNITest.cs

```diff
@@ -0,0 +1,125 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Net.Test.Common;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Security.Tests
+{
+    using Configuration = System.Net.Test.Common.Configuration;
+
+    [Trait("feature", "sni")]
+    public class SslStreamSNITest
+    {
+        private static IEnumerable<object[]> HostNameData()
+        {
+            yield return new object[] { "a" };
+            yield return new object[] { "test" };
+            yield return new object[] { new string('a', 100) };
+        }
+
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            WithVirtualConnection((server, client) =>
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+
+                int timesCallbackCalled = 0;
+                options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
+                {
+                    timesCallbackCalled++;
+                    Assert.Equal(hostName, actualHostName);
+                    return serverCert;
+                };
+
+                var cts = new CancellationTokenSource();
+                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
+
+                Assert.Equal(1, timesCallbackCalled);
+                clientJob.Wait();
+            },
+            (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+                {
+                    Assert.Equal(serverCert, certificate);
+                    return true;
+                });
+        }
+
+        [Fact]
+        public void SslStream_UnknownHostName_Fails()
+        {
+            WithVirtualConnection((server, client) =>
+            {
+                Task clientJob = Task.Run(() => {
+                    Assert.Throws<VirtualNetwork.VirtualNetworkConnectionBroken>(()
+                        => client.AuthenticateAsClient("test"));
+                });
+
+                int timesCallbackCalled = 0;
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+                options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
+                {
+                    timesCallbackCalled++;
+                    return null;
+                };
+
+                var cts = new CancellationTokenSource();
+                Assert.ThrowsAsync<NotSupportedException>(async () => {
```

Comment:
Open question is what should we throw when callback returns null - currently there is a code handling null case and I have not added anything new but this is something we might consider changing.

Reply:
What does SslStream throw today if you try to AuthenticateAsServerAsync with a null certificate? whatever that does is probably going to make most sense?

Reply:
this scenario is being detected early - in this case we don't know about null until we receive hello from the client

Reply:
We need a test for this to make sure that the exception we might throw if catch-able by the developer's code. We don't want, for instance, to throw an exception on a threadpool thread.

Reply:
@davidsh this is a test - I will likely change the exception once I figure out which one is the most appropriate here.